corehq/apps/app_manager/fields.py

```diff
@@ -116,3 +139,373 @@ def get_app_sources(domain):
+            if module:
+                # module is sometimes wrapped json, sometimes a dict!
+                module_id = module['id'] if 'id' in module else module.id
+            else:
+                module_id = -1 if form.get('is_user_registration') else 1000
+            app_form = form.get('form')
+            if app_form:
+                # app_form is sometimes wrapped json, sometimes a dict!
+                form_id = app_form['id'] if 'id' in app_form else app_form.id
+            else:
+                form_id = -1
+            return (order, app_name, app_id, module_id, form_id)
+        else:
+            form_xmlns = form['xmlns']
+            return (2, form_xmlns, app_id)
+
+    @property
+    @memoized
+    def _all_forms(self):
+        forms = []
+        unknown_forms = []
+        for f in get_exports_by_form(self.domain):
+            form = f['value']
+            if form.get('app_deleted') and not form.get('submissions'):
+                continue
+            if 'app' in form:
+                form['has_app'] = True
+                forms.append(form)
+            else:
+                app_id = f['key'][1] or ''
+                form['app'] = {
+                    'id': app_id
+                }
+                form['has_app'] = False
+                form['show_xmlns'] = True
+                unknown_forms.append(form)
+        forms.extend(self._attach_unknown_suggestions(unknown_forms))
+        return sorted(forms, key=self._sorkey_form)
+
+    @property
+    @memoized
+    def _no_app_forms(self):
+        return filter(lambda f: not f.get('has_app', False), self._all_forms)
+
+    @property
+    @memoized
+    def _remote_app_forms(self):
+        return filter(lambda f: f.get('has_app', False) and f.get('show_xmlns', False), self._all_forms)
+
+    @property
+    @memoized
+    def _deleted_app_forms(self):
+        return filter(
+            lambda f: f.get('has_app', False) and f.get('app_deleted') and not f.get('show_xmlns', False),
+            self._all_forms
+        )
+
+    @property
+    @memoized
+    def _available_app_forms(self):
+        return filter(
+            lambda f: f.get('has_app', False) and not f.get('app_deleted') and not f.get('show_xmlns', False),
+            self._all_forms
+        )
+
+    def _get_app_type_choices(self, as_dict=True):
+        choices = [(_("All Applications"), self.APP_TYPE_ALL)]
+        if self._remote_app_forms:
+            choices.append((_("Remote Application"), self.APP_TYPE_REMOTE))
+        if self._deleted_app_forms:
+            choices.append((_("Deleted Application"), self.APP_TYPE_DELETED))
+        if self._no_app_forms:
+            choices.append((_("No Application"), self.APP_TYPE_NONE))
+        choices = map(lambda c: RMIDataChoice(id=c[1], text=c[0], data={}), choices)
+        if as_dict:
+            choices = map(lambda c: c._asdict(), choices)
+        return choices
+
+    @staticmethod
+    def _get_unique_choices(choices):
```

Comment:
pretty hard to parse this since these are vague names and unclear data structures.